src/data/data.cc

```diff
@@ -47,19 +48,20 @@ void MetaInfo::SaveBinary(dmlc::Stream *fo) const {
 }
 
 void MetaInfo::LoadBinary(dmlc::Stream *fi) {
-  int version;
-  CHECK(fi->Read(&version, sizeof(version)) == sizeof(version)) << "MetaInfo: invalid version";
-  CHECK(version >= 1 && version <= kVersion) << "MetaInfo: unsupported file version";
+  auto version = Version::Load(fi);
+  auto major = std::get<0>(version);
+  // MetaInfo is saved in `SparsePageSource'.  So the version in MetaInfo represents the
+  // version of DMatrix.
+  CHECK_EQ(major, 1) << "Binary DMatrix generated by XGBoost: "
```

Comment:
What about loading from future versions? The statement means we can only load from the same major version. Maybe this makes sense I'm not sure. This error message would also be wrong if we loaded from a hypothetical major version 2.

Reply:
Just trying to avoid checking the version saved before this PR (which is 3).  It's a little bit awkward, since the error message doesn't make sense at all for version before this PR.  The version string might be complete nonsense considering that old binary file saves only 1 integer and we are now reading 3 of them.  One possible workaround is to prefix with a string like "version:", then we know whether the binary is saved before 1.0.

Reply:
Or, we can save the version as string instead of integers.

Reply:
@RAMitchell I now write the string "version:" into binary file.